In [1]:
import pandas as pd
import numpy as np

In [2]:
google_places_path = "/home/jczestochowska/workspace/ada/chicago_food/data/processed/google_places.csv"
restaurants_path = "/home/jczestochowska/workspace/ada/chicago_food/data/processed/restaurants.csv"

In [3]:
google_places = pd.read_csv(google_places_path)
restaurants = pd.read_csv(restaurants_path)

Let's see if all restaurants we got from google are the ones we asked for and are in Chicago

In [4]:
google_places.head()

place_id             place_name  rating  \
0  ChIJleWegJHTD4gRsIMGm3IjUcM    yolk - test kitchen     4.4   
1  ChIJI1dehI0xDogRRkvKIpUspHM  las asadas restaurant     4.0   
2  ChIJBRuLGC_MD4gRfUVkzaQBIZk          mini palapita     4.3   
3  ChIJ-auSyWItDogRQwUqcwljejE     crazy bird chicken     4.7   
4  ChIJIYRFg7IsDogRAW0zLguKL-A      redhead piano bar     4.2   

   total_number_of_ratings  price_level               address     city  \
0                    180.0          2.0  1767 n milwaukee ave  chicago   
1                    166.0          2.0        3834 w 47th st  chicago   
2                    342.0          1.0     4968 n elston ave  chicago   
3                     67.0          NaN      1160 w grand ave  chicago   
4                    989.0          2.0       16 w ontario st  chicago   

     zip_code  
0       60647  
1  60632-4136  
2       60630  
3  60642-5837  
4       60654

In [5]:
restaurants.head()

place_id                place_name   latitude  \
0  ChIJleWegJHTD4gRsIMGm3IjUcM         yolk test kitchen  41.913588   
1  ChIJI1dehI0xDogRRkvKIpUspHM  las asadas mexican grill  41.808025   
2  ChIJBRuLGC_MD4gRfUVkzaQBIZk               la palapita  41.808025   
3  ChIJ-auSyWItDogRQwUqcwljejE                crazy bird  41.891193   
4  ChIJIYRFg7IsDogRAW0zLguKL-A     the redhead piano bar  41.893371   

   longitude                address  zip_code  
0 -87.682203  1767 n milwaukee ave    60647.0  
1 -87.720037        3834 w 47th st    60632.0  
2 -87.720037        3834 w 47th st    60632.0  
3 -87.657055      1160 w grand ave    60642.0  
4 -87.628783    16-18 w ontario st    60654.0

We notice that address from chicago database has whitespace at the end so we remove it from each row

In [6]:
restaurants.address = restaurants.address.apply(str.strip)

In [7]:
print(f"Google places API returned {len(google_places)} restaurants")

Google places API returned 13006 restaurants


Check if all restaurants are in Chicago

In [8]:
not_chicago_restaurants = google_places[~(google_places.city == "chicago")]
print(f"Number of restaurants with non-chicago city {len(not_chicago_restaurants)}")
not_chicago_restaurants.city.unique()

Number of restaurants with non-chicago city 2053


array(['60666', 'des plaines', 'libertyville', 'asheville', 'schaumburg',
       'elk grove village', 'morton grove', 'wauwatosa', '60622',
       'oak park', 'harwood heights', 'frankfort', 'lubbock', 'houston',
       '60641', 'holton', '45014', nan, 'naperville', '28594', 'rosemont',
       'carol stream', 'stone park', 'westchester', '46617', 'cicero',
       'forest park', 'madison', 'crestwood', 'addison', 'mancelona',
       'khet phra nakhon', '87544', 'sterling heights', '60642',
       'downers grove', 'st. louis', '050015', '60618', 'cary', 'joliet',
       'east chicago', 'miami', 'roseland', '60632', 'new york', '60603',
       'wilmette', '60649', 'dallas', 'evanston', '60623', 'st. charles',
       'bolingbrook', 'hoffman estates', 'springfield', 'elgin', '1205',
       'atlanta', '60629', 'detroit', '60657', 'skokie', 'wood dale',
       'thornlie', 'hickory hills', 'clayton', 'hinsdale', 'highland',
       'franklin park', 'yuma', 'lansing', 'wheaton', 'bellwood',
    

We can see that sometimes the zip code was returned as a city, Chicago zip codes start with 60 therefore those entries where city equals a number starting 60 are probably in Chicago and their city name is in address column

In [9]:
address_chicago_count = len(google_places[google_places.address == "chicago"])
print(f"Number of places with chicago as address and zip code in city column is {address_chicago_count}")

Number of places with chicago as address and zip code in city column is 163


In [10]:
google_places[google_places.city.str.startswith("60", na=False)]

place_id                          place_name  \
13     ChIJVwHOlyi0D4gRNjLVMUoK00A             romano's macaroni grill   
74     ChIJqQyWfSa0D4gRN_IZ45o_HCE                            argo tea   
106    ChIJTxSyOsfSD4gRUy8JoHTgkVw                           piece out   
183    ChIJIVrbHPjMD4gRQfUHdZXic5U                  la estrella blanca   
278    ChIJM_1zjXe2D4gRDAmIxvGytuw                 chili's grill & bar   
...                            ...                                 ...   
12644  ChIJq9zsX_TSD4gRrRW0u2HLN1E                   sweet bean & more   
12699  ChIJj_2hl5DMD4gRNunA9w1SI1I  central starlite family restaurant   
12775  ChIJz5h8elIzDogRgDuu28QU6Ic                    ruby's soul food   
12813  ChIJDcM0Ey4zDogR2rNIPGIV_gs                       come & get it   
12957  ChIJ3aY5ZpsvDogRlumoNGkp9Xk               amanda j unisex salon   

       rating  total_number_of_ratings  price_level  address   city zip_code  
13        3.7                    860.0          2.0  chicago  60666      NaN  
74        3.5                     45.0          1.0  chicago  60666      NaN  
106       NaN                      NaN          NaN  chicago  60622      NaN  
183       3.9                     29.0          NaN  chicago  60641      NaN  
278       3.2                    913.0          2.0  chicago  60666      NaN  
...       ...                      ...          ...      ...    ...      ...  
12644     5.0                      2.0          NaN  chicago  60614      NaN  
12699     NaN                      NaN          NaN  chicago  60639      NaN  
12775     NaN                      NaN          NaN  chicago  60651      NaN  
12813     NaN                      NaN          NaN  chicago  60644      NaN  
12957     NaN                      NaN          NaN  chicago  60620      NaN  

[173 rows x 8 columns]

If zip code starts with 60 move it to zip code column

In [11]:
chicago_idx = google_places.city.str.startswith("60", na=False)
google_places.loc[chicago_idx, 'zip_code'] = google_places['city']

In [12]:
google_places[google_places.city.str.startswith("60", na=False)]

place_id                          place_name  \
13     ChIJVwHOlyi0D4gRNjLVMUoK00A             romano's macaroni grill   
74     ChIJqQyWfSa0D4gRN_IZ45o_HCE                            argo tea   
106    ChIJTxSyOsfSD4gRUy8JoHTgkVw                           piece out   
183    ChIJIVrbHPjMD4gRQfUHdZXic5U                  la estrella blanca   
278    ChIJM_1zjXe2D4gRDAmIxvGytuw                 chili's grill & bar   
...                            ...                                 ...   
12644  ChIJq9zsX_TSD4gRrRW0u2HLN1E                   sweet bean & more   
12699  ChIJj_2hl5DMD4gRNunA9w1SI1I  central starlite family restaurant   
12775  ChIJz5h8elIzDogRgDuu28QU6Ic                    ruby's soul food   
12813  ChIJDcM0Ey4zDogR2rNIPGIV_gs                       come & get it   
12957  ChIJ3aY5ZpsvDogRlumoNGkp9Xk               amanda j unisex salon   

       rating  total_number_of_ratings  price_level  address   city zip_code  
13        3.7                    860.0          2.0  chicago  60666    60666  
74        3.5                     45.0          1.0  chicago  60666    60666  
106       NaN                      NaN          NaN  chicago  60622    60622  
183       3.9                     29.0          NaN  chicago  60641    60641  
278       3.2                    913.0          2.0  chicago  60666    60666  
...       ...                      ...          ...      ...    ...      ...  
12644     5.0                      2.0          NaN  chicago  60614    60614  
12699     NaN                      NaN          NaN  chicago  60639    60639  
12775     NaN                      NaN          NaN  chicago  60651    60651  
12813     NaN                      NaN          NaN  chicago  60644    60644  
12957     NaN                      NaN          NaN  chicago  60620    60620  

[173 rows x 8 columns]

if address is chicago move it to city column

In [13]:
address_idx = (google_places.address == "chicago")
google_places.loc[address_idx, "city"] = "chicago"
google_places.loc[address_idx, "address"] = np.nan

In [14]:
google_places[google_places.address == 'chicago']

Empty DataFrame
Columns: [place_id, place_name, rating, total_number_of_ratings, price_level, address, city, zip_code]
Index: []

In [15]:
not_chicago_restaurants = google_places[~(google_places.city == "chicago")]
print(f"Number of restaurants with non-chicago city {len(not_chicago_restaurants)}")
not_chicago_restaurants.city.unique()

Number of restaurants with non-chicago city 1890


array(['des plaines', 'libertyville', 'asheville', 'schaumburg',
       'elk grove village', 'morton grove', 'wauwatosa', 'oak park',
       'harwood heights', 'frankfort', 'lubbock', 'houston', 'holton',
       '45014', nan, 'naperville', '28594', 'rosemont', 'carol stream',
       'stone park', 'westchester', '46617', 'cicero', 'forest park',
       'madison', 'crestwood', 'addison', 'mancelona', 'khet phra nakhon',
       '87544', 'sterling heights', 'downers grove', 'st. louis',
       '050015', 'cary', 'joliet', 'east chicago', 'miami', 'roseland',
       'new york', 'wilmette', 'dallas', 'evanston', 'st. charles',
       'bolingbrook', 'hoffman estates', 'springfield', 'elgin', '1205',
       'atlanta', 'detroit', 'skokie', 'wood dale', 'thornlie',
       'hickory hills', 'clayton', 'hinsdale', 'highland',
       'franklin park', 'yuma', 'lansing', 'wheaton', 'bellwood',
       'north york', 'grand rapids', 'milwaukee', 'bridgeview', 'jeddah',
       'oakbrook terrace', 'hillside

We can see that sometimes 60 startin zip codes are still present, let check city for those values

In [16]:
google_places[google_places.city.str.startswith("60", na=False)]

place_id                           place_name  \
2408   ChIJ6xFHBL6vD4gRchf5OrA9qUg      northwest transportation center   
5469   ChIJsTHl8TevD4gRVlyTfs6j5Y0              nic's organic fast food   
6613   ChIJzQ0X9bYzDogRHT07KfAcfNA                              mc café   
7438   ChIJT0JHiqUzDogRlqHkJuSCgrs                           nueva vida   
8272   ChIJt1juNU80DogRzGgeK6nsrs0  red robin gourmet burgers and brews   
8551   ChIJo3gfbs66D4gRNGDpJP0lFEo            mr d's sports bar & grill   
8572   ChIJR5CMqiGuD4gR3N5M6VH7ryE                vincenzo palmieri dpm   
9575   ChIJQ-X4Dq4zDogR543sVOJGvsk               hibachi seafood buffet   
9736   ChIJ7aMROx4FD4gRRAVVecrnbUY                fantasea destinations   
9794   ChIJs-zK38JKDogROyQY0tbyclU                        sharky's fish   
9878   ChIJEU3bws6wD4gRc-_2U0GQAwI                        maraca's club   
10122  ChIJT0JHiqUzDogRlqHkJuSCgrs                           nueva vida   
10162  ChIJDXukEK4zDogRoL9f3mLUeFc          my twirl yogurt & smoothies   
10818  ChIJ3bHi-DbOD4gRcGXLyROZV-s                        gfc buffalo's   
10935  ChIJJajT_VMrDogRhxDxFgORep8                           ogden slip   
11686  ChIJkXTFui71DogR2MaMn2rp1t0                smiley's gyros & beef   
12005  ChIJtUj3XHsjDogRqzYmJikzjTc                birria estilo ocotlan   

       rating  total_number_of_ratings  price_level            address   city  \
2408      4.0                     20.0          NaN         schaumburg  60173   
5469      3.1                     31.0          NaN         schaumburg  60173   
6613      4.5                      2.0          NaN             cicero  60804   
7438      4.2                      6.0          NaN             cicero  60804   
8272      3.9                   1032.0          2.0    north riverside  60546   
8551      4.3                     56.0          NaN  arlington heights  60005   
8572      4.8                     19.0          NaN  elk grove village  60007   
9575      3.4                      8.0          NaN             cicero  60804   
9736      NaN                      NaN          NaN              elgin  60123   
9794      4.5                      2.0          NaN           bellwood  60104   
9878      4.0                      1.0          NaN        des plaines  60016   
10122     4.2                      6.0          NaN             cicero  60804   
10162     4.6                      8.0          NaN             cicero  60804   
10818     3.6                     18.0          NaN        lincolnwood  60712   
10935     4.3                      3.0          NaN      streeterville  60611   
11686     4.7                    353.0          1.0         plainfield  60586   
12005     4.4                      7.0          NaN        blue island  60406   

      zip_code  
2408     60173  
5469     60173  
6613     60804  
7438     60804  
8272     60546  
8551     60005  
8572     60007  
9575     60804  
9736     60123  
9794     60104  
9878     60016  
10122    60804  
10162    60804  
10818    60712  
10935    60611  
11686    60586  
12005    60406

It turns out that entries as cicero, north riverside, schaumburg or elgin are within Chicago aglomeration they are smaller towns or districts.

Let's check how many entries don't have Chicago area zip code

In [17]:
google_places[(~(google_places.zip_code.str.startswith("60", na=False)) & (google_places.city != "chicago"))]

place_id                     place_name  rating  \
25     ChIJ8W5KpKj0WYgRXlQU3Lknu-c  liberty house coffee and café     4.7   
94     ChIJs1HOuK4FBYgREBfd2sHyGwg                     la terraza     4.6   
163    ChIJFXchGjxt_oYR3OZ7Dv4sz3A            mi linda michoacana     4.5   
174    ChIJ2TzjVki8QIYR8rBeeXUQuGA          doña chela restaurant     4.3   
186    ChIJf4UCjadCGYgRPk63PEDsfwo          compass group usa inc     NaN   
...                            ...                            ...     ...   
12967  ChIJ8WUga24zXIgRrtJ5T0sC08M          mediterranean delight     4.8   
12974  ChIJScl8s92LOIgRwRKOCn4btuI                 couscous house     4.9   
12975  ChIJrUpLxVkDBYgR2GJViXvlxkY               champion chicken     3.5   
13001  ChIJIVSzGQ3nEYgR8nZhwaBQ9ZM         food king china buffet     3.6   
13003  ChIJtf1ZQfzqEYgR9gh_2CPqkNE                cagney's saloon     4.0   

       total_number_of_ratings  price_level                   address  \
25                       234.0          2.0          221 s liberty st   
94                        77.0          NaN  11520 w bluemound rd # 2   
163                      240.0          NaN              2002 34th st   
174                      212.0          1.0              1112 76th st   
186                        NaN          NaN             6477 syers rd   
...                        ...          ...                       ...   
12967                    237.0          1.0     160 bus terminal rd a   
12974                    100.0          1.0             1611 morse rd   
12975                    391.0          1.0         8718 w lisbon ave   
13001                    224.0          1.0            214 w ridge rd   
13003                    380.0          1.0               236 main st   

            city    zip_code  
25     asheville  28801-2334  
94     wauwatosa  53226-4000  
163      lubbock  79411-1832  
174      houston  77012-1004  
186       holton  49425-7508  
...          ...         ...  
12967  oak ridge       37830  
12974   columbus       43229  
12975  milwaukee  53222-2859  
13001   griffith  46319-1041  
13003     hobart       46342  

[1071 rows x 8 columns]

We have 1071 rows which correspond to restaurants in different places than Chicago area. They are of no use to us so we remove them.

In [18]:
google_places_cleaned = google_places[((google_places.zip_code.str.startswith("60", na=False)) | (google_places.city == "chicago"))]

In [19]:
google_places_cleaned

place_id               place_name  rating  \
0      ChIJleWegJHTD4gRsIMGm3IjUcM      yolk - test kitchen     4.4   
1      ChIJI1dehI0xDogRRkvKIpUspHM    las asadas restaurant     4.0   
2      ChIJBRuLGC_MD4gRfUVkzaQBIZk            mini palapita     4.3   
3      ChIJ-auSyWItDogRQwUqcwljejE       crazy bird chicken     4.7   
4      ChIJIYRFg7IsDogRAW0zLguKL-A        redhead piano bar     4.2   
...                            ...                      ...     ...   
12999  ChIJW9Fy4mXLD4gRZa_kB9Ch0y8  golden crown restaurant     4.6   
13000  ChIJF1jwR_AsDogRqxHTjLkRiCk        lalo's on maxwell     3.9   
13002  ChIJ5fNghEVMDogR4UWYi3iHLPE          dip'n good dips     NaN   
13004  ChIJ6a1jjZssDogR3PQwtmza-vs            mei's kitchen     3.5   
13005  ChIJubeVEQrTD4gRVKoy9DuCqsQ          surf associates     NaN   

       total_number_of_ratings  price_level                  address  \
0                        180.0          2.0     1767 n milwaukee ave   
1                        166.0          2.0           3834 w 47th st   
2                        342.0          1.0        4968 n elston ave   
3                         67.0          NaN         1160 w grand ave   
4                        989.0          2.0          16 w ontario st   
...                        ...          ...                      ...   
12999                     11.0          NaN  2341 n narragansett ave   
13000                   1005.0          2.0         733 w maxwell st   
13002                      NaN          NaN          17 monterey ave   
13004                     86.0          NaN      1108 s michigan ave   
13005                      NaN          NaN            528 w surf st   

             city    zip_code  
0         chicago       60647  
1         chicago  60632-4136  
2         chicago       60630  
3         chicago  60642-5837  
4         chicago       60654  
...           ...         ...  
12999     chicago  60639-2654  
13000     chicago  60607-5016  
13002  villa park  60181-2822  
13004     chicago       60605  
13005     chicago  60657-6066  

[11935 rows x 8 columns]

We will remove any duplicating rows from both dataframes

In [20]:
google_places_cleaned.drop_duplicates(inplace=True)

/home/jczestochowska/anaconda3/envs/ada/lib/python3.7/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [21]:
restaurants.drop_duplicates(inplace=True)

Now we want to compare addresses and names of restaurants. we will start by joining two dataframes on place_id

In [22]:
merged = restaurants.merge(google_places_cleaned, on="place_id", how="inner", suffixes=("_restaurants","_google"))

In [23]:
len(merged)

12057

In [24]:
merged.columns

Index(['place_id', 'place_name_restaurants', 'latitude', 'longitude',
       'address_restaurants', 'zip_code_restaurants', 'place_name_google',
       'rating', 'total_number_of_ratings', 'price_level', 'address_google',
       'city', 'zip_code_google'],
      dtype='object')

to assess if for sure we have correct entries we will compare zip codes if no zip_code is available from restaurants then we will compare names, first step would be to check where `zip_code_restaurants` is NaN
and second would be to remove second zip code component from `zip_code_google`

In [25]:
merged[merged.zip_code_restaurants.isna()]

place_id  place_name_restaurants   latitude  \
960    ChIJq7u58SPTD4gRqV7j2UThRLM      s.g.a. coffee cafe  41.958548   
1351   ChIJo3LREajSD4gROdM5Sjl8pTg  casa central la posada  41.878100   
7752   ChIJBV-jIEvTD4gROrQeGm3dLt0       starbucks coffee   41.896318   
8872   ChIJV1gyFIErDogR4vUnQw74VN4         mccornick place  41.851047   
8873   ChIJV1gyFIErDogR4vUnQw74VN4         mccornick place  41.851047   
10790  ChIJTwjQ-HssDogRz558Go2FZu8            burger king   41.850454   
11136  ChIJYZbHo0zTD4gReP8dX6Qfaco      think simple foods  41.896504   

       longitude       address_restaurants  zip_code_restaurants  \
960   -87.786919       4300 n narragansett                   NaN   
1351         NaN                                             NaN   
7752  -87.635840        750 n franklin st.                   NaN   
8872  -87.622093            2301 s indiana                   NaN   
8873  -87.622093            2301 s indiana                   NaN   
10790 -87.623891  2320-2322 s michigan ave                   NaN   
11136 -87.632784         141 w chicago ave                   NaN   

           place_name_google  rating  total_number_of_ratings  price_level  \
960    coffee lab & roasters     4.9                     40.0          NaN   
1351            casa central     3.5                     31.0          NaN   
7752               starbucks     4.0                    275.0          2.0   
8872         mccormick place     4.5                   8493.0          NaN   
8873         mccormick place     4.5                   8494.0          NaN   
10790            burger king     3.4                    731.0          1.0   
11136           think simple     5.0                      2.0          NaN   

              address_google     city zip_code_google  
960       2823 n lincoln ave  chicago           60657  
1351   1343 n california ave  chicago      60622-2803  
7752       750 n franklin st  chicago           60654  
8872          2301 s king dr  chicago           60616  
8873          2301 s king dr  chicago           60616  
10790    2328 s michigan ave  chicago           60616  
11136       770 n lasalle dr  chicago           60610

We have only couple of rows where zip_code_restaurants is not available, we can review them manually, it turns out we will have to remove row 960, 11136 and McCormick place which is not a restaurant but a business center in Chicago.

In [26]:
merged.drop(axis=0, index=[960, 11136, 8872,8873], inplace=True)

Now we will remove second zip code google component

In [27]:
merged[merged.zip_code_google.isna()]

place_id            place_name_restaurants  \
2974   ChIJQQMfpi_TD4gRbkamIyMmFvU                  kikka- kingsbury   
3411   ChIJ2aXLDIA3DogRtia7eMEyzE4  harris family grill on 83rd, llc   
3751   ChIJ-fxWG6PND4gR7_Y1lx9ULuo                 elston nugget inc   
6132   ChIJedCe4RXOD4gRl2sn7ZmQxyQ                       sushi world   
6133   ChIJedCe4RXOD4gRl2sn7ZmQxyQ                        sushi luxe   
6702   ChIJrS7zoBzND4gRFlZJQaAkBBw                           hermosa   
9671   ChIJra52C_AsDogRw9Xe4NvtHps       maxwell st & fresh lemonade   
9672   ChIJra52C_AsDogRw9Xe4NvtHps                    maxwell street   
10021  ChIJ1XHQCS4tDogRehV3FO2Qr_M                          grandbar   
10468  ChIJqRc0MtMoDogR6i4WiN_xSk0                    new park manor   
10698  ChIJ7cv00DwsDogRAMDACa2m4K8                          chi-town   
10699  ChIJ7cv00DwsDogRAMDACa2m4K8         pr cafeteria incorporated   
11005  ChIJHcxIkUvTD4gRv6-Fg_LXPP4                all stars on wells   
11883  ChIJn6M0ivwxDogR4z5ls6I-waA                cody's  on  albany   
12047  ChIJB3LyNm4wDogRg_HvhImIlNQ             79th  & state brown's   

        latitude  longitude                 address_restaurants  \
2974   41.909399 -87.653051                 1550 n kingsbury st   
3411   41.744483 -87.556373                      2801 e 83rd st   
3751   41.932281 -87.688190                 2406 w diversey ave   
6132   41.995479 -87.712875                 6251 n mccormick rd   
6133   41.976357 -87.668568                     5204 n clark st   
6702   41.917156 -87.736189                 4356 w armitage ave   
9671   41.794426 -87.644363                 740 w garfield blvd   
9672   41.909670 -87.751213                    4948 w north ave   
10021  41.891087 -87.667204                    1600 w grand ave   
10468  41.765395 -87.615110  7109 s dr martin luther king jr dr   
10698  41.883172 -87.652133              1000 w washington blvd   
10699  41.909898 -87.707145                   1552 n kedzie ave   
11005  41.886707 -87.634150                     205 w wacker dr   
11883  41.954255 -87.705437              4009 n albany ave bldg   
12047  41.751010 -87.624401                     7859 s state st   

       zip_code_restaurants       place_name_google  rating  \
2974                60642.0  north kingsbury street     NaN   
3411                60617.0        west 83rd street     NaN   
3751                60647.0     north elston avenue     NaN   
6132                60659.0             sushi world     4.4   
6133                60640.0             sushi world     4.4   
6702                60639.0                 hermosa     NaN   
9671                60609.0     west maxwell street     NaN   
9672                60639.0     west maxwell street     NaN   
10021               60622.0         grand & ashland     NaN   
10468               60619.0              park manor     NaN   
10698               60607.0                 chicago     NaN   
10699               60651.0                 chicago     NaN   
11005               60606.0      north wells street     NaN   
11883               60618.0     south albany avenue     NaN   
12047               60619.0        west 79th street     NaN   

       total_number_of_ratings  price_level  address_google     city  \
2974                       NaN          NaN  n kingsbury st  chicago   
3411                       NaN          NaN       w 83rd st  chicago   
3751                       NaN          NaN    n elston ave  chicago   
6132                     410.0          2.0             NaN  chicago   
6133                     410.0          2.0             NaN  chicago   
6702                       NaN          NaN             NaN  chicago   
9671                       NaN          NaN    w maxwell st  chicago   
9672                       NaN          NaN    w maxwell st  chicago   
10021                      NaN          NaN             NaN  chicago   
10468                      NaN          NaN             NaN  chicago   

Most of those entries also have rating NaN, drop then as those rows are useless for us

In [28]:
merged.dropna(subset=["rating"],axis=0, inplace=True)

Filling nans with string value to remove second `zip_code_google` component

In [29]:
merged.zip_code_google.fillna("nan", inplace=True)

In [30]:
merged.zip_code_google.str.len().unique()

array([ 5, 10,  3])

In [31]:
merged.zip_code_google = merged.zip_code_google.apply(lambda x: x[:5])

In [32]:
merged.zip_code_google = merged.zip_code_google.replace("nan", np.nan)

In [33]:
merged.zip_code_google = merged.zip_code_google.apply(float)

Now we will take only those entries where every place has a matching zip code

In [34]:
matching = merged[merged.zip_code_google == merged.zip_code_restaurants]

At last to be sure that we have rating corresponding to a correct place we will compare names using python difflib, from docs "ratio() returns a float in [0, 1], measuring the similarity of the sequences. As a rule of thumb, a ratio() value over 0.6 means the sequences are close matches:". We will use this property to discard any rows where name similarity is lower than 0.6

In [35]:
from difflib import SequenceMatcher

def get_similarity_score(restaurant_name, google_name):
    return SequenceMatcher(lambda x: x == " ",
                    restaurant_name,
                    google_name).ratio()

In [36]:
matching['name_similarity_score'] = matching.apply(lambda row: get_similarity_score(row["place_name_restaurants"], row["place_name_google"]), axis=1)

/home/jczestochowska/anaconda3/envs/ada/lib/python3.7/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [37]:
matching = matching[matching.name_similarity_score > 0.6]

As a last step we remove rows with duplicated place_id

In [38]:
matching = matching.drop_duplicates(subset=["place_id"])

In [39]:
len(matching)

5700

we have ratings for 5700 restaurants

We will transform matching dataframe to use it for further analysis, we will remove useless columns and save it to csv file

In [40]:
matching.drop(["name_similarity_score", "place_name_google", "price_level", "address_google", "zip_code_google", "city"], axis=1, inplace=True)

In [41]:
matching.rename(columns={"place_name_restaurants": "place_name", "address_restaurants": "address", "zip_code_restaurants": "zip_code"}, inplace=True)

In [43]:
matching.to_csv("../data/processed/restaurants_with_ratings.csv")